src/System.Net.HttpListener/tests/HttpListenerContextTests.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Net.Sockets;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Tests
+{
+    public class HttpListenerContextTests
+    {
+        public static bool IsNotWindows7 { get; } = !PlatformDetection.IsWindows7 && PlatformDetection.IsNotOneCoreUAP;
+
+        public static IEnumerable<object[]> SubProtocol_TestData()
+        {
+            yield return new object[] { new string[0], null };
+
+            yield return new object[] { new string[] { "MyProtocol1" }, null };
+            yield return new object[] { new string[] { "MyProtocol" }, "MyProtocol" };
+            yield return new object[] { new string[] { "MyProtocol" }, "myPROTOCOL" };
+
+            yield return new object[] { new string[] { "MyProtocol1", "MyProtocol2" }, null };
+            yield return new object[] { new string[] { "MyProtocol1", "MyProtocol2" }, "MyProtocol2" };
+        }
+
+        [ConditionalTheory(nameof(IsNotWindows7))]
+        [MemberData(nameof(SubProtocol_TestData))]
+        public async Task AcceptWebSocketAsync_ValidSubProtocol_Success(string[] clientProtocols, string serverProtocol)
+        {
+            await GetWebSocketContext((socket, context) =>
+            {
+                HttpListenerWebSocketContext socketContext = context.AcceptWebSocketAsync(serverProtocol).Result;
+                Assert.Equal(serverProtocol, socketContext.WebSocket.SubProtocol);
+            }, subProtocols: clientProtocols);
+        }
+
+        [ConditionalFact(nameof(IsNotWindows7))]
+        // The managed implementation doesn't validate that the socket is actually a web socket.
+        // This means that HttpListener can connect to a Socket masquerading as a web socket.
+        [ActiveIssue(18128, TestPlatforms.AnyUnix)]
+        public async Task AcceptWebSocketAsync_SocketSpoofingAsWebSocket_ThrowsWebSocketException()
```

Comment:
I don't understand the "masquerading as a web socket" part.  Web sockets are just layered on top of http, which are implemented on sockets, so every "web socket" is a socket.  It sounds like you're saying that the managed implementation simply isn't validating all of the headers the way it should, in which case that's something we should fix.